src/prototype_creep_routing.js

```diff
@@ -177,27 +177,31 @@ Creep.prototype.followPath = function(action) {
     return action(this);
   }
   const prepareData = this.moveByPathPrepare(route, routePos, 'pathStart', this.memory.routing.targetId);
-  if (prepareData.unit.preMove) {
-    if (prepareData.unit.preMove(this, prepareData.directions)) {
-      return true;
-    }
+  if (prepareData !== false) {
+    return this.moveByPathMy(route, routePos, 'pathStart', this.memory.routing.targetId, action, prepareData);
   }
-  return this.moveByPathMy(route, routePos, 'pathStart', this.memory.routing.targetId, action, prepareData);
+  return true;
 };
 
 Creep.prototype.moveByPathPrepare = function(route, routePos, start, target) {
   const result = {};
   result.unit = roles[this.memory.role];
+
   // Somehow reset the pathPos if the path has changed?!
   result.path = this.room.getPath(route, routePos, start, target);
-  if (!result.path) {
-    return result;
+  if (result.path) {
+    result.pathPos = this.getPathPos(route, routePos, result.path);
+    if (result.pathPos > -1) {
+      result.directions = this.getDirections(result.path, result.pathPos);
+    }
   }
-  result.pathPos = this.getPathPos(route, routePos, result.path);
-  if (result.pathPos < 0) {
-    return result;
+  if ((result.unit.preMove && result.unit.preMove(this, result.directions)) || !result.path) {
```

Comment:
we mustn't call preMove here, because this may potentially lead to infinite recursion:
harvester.preMove->...->moveByPathMy->moveByPathPrepare->harvester.preMove
this was the reason of extracting moveByPathPrepare

Reply:
preMove shouldn't call moveByPathMY
if a role do it then it is that which have to be fixed

Reply:
even if it never goes into recursion now because of some conditions, we should only call preMove from followPath, to have neat call  graph in recursion, and to not fail into recursion if we'd accidentally break conditions that prevent it

Reply:
@Nevrdid, maybe, but it is called now. there was `noPreMove` flag, before `moveByPathPrepare` extraction

Reply:
ok will look at how to export it

Reply:
#394

Reply:
nopremove flag ?

Reply:
yeah seen it else i shouldn't say it come from here :)  don't you see these preMove you just removed?

Reply:
yeah, that was `skipPreMove`
I suppose that I've failed and introduce regression while refactoring, but I still don't see where is it)

Reply:
alright, exported checks in a new prototype. do you seen betetr the difference ?

Reply:
nope. 😞 
I've figured out that I had messed up with order of calling `preMove` and setting `memory.routing.pathPos`, but it not looks like this it the thing you've fixed. All other conditional branches, as far as I can see will lead to the same result. I mean that they shouldn't be the same, as you've fixed something, but I can't figure out what it was)
Maybe you'd just describe what branch if failed in my change?)

Reply:
Ah! I found something... After you change we return `true` instead of `false` in case of `datas.pathPos < 0 && datas.path.length === 0`. But we haven't this behavior before my change)

Reply:
yeah ^^' sorry didn't respond as i fixing it again ^^' as simplifying it i broke it again x/

Reply:
the problem is followPath returns which differ depending prepare datas. but we can't really do a returnFalse() method as we have to altern differents checks.  will push back soon.

Reply:
alright ^^ fixed.

Reply:
makes more sense now. so setting `memory.routing.pathPos` before `preMove` was really the case.